Site/SiteConfigModule.php

```diff
@@ -91,685 +91,659 @@
+     *
+     * @param string $section the config setting section
+     * @param string $section the config setting name
+     * @param int    $source  either {@link SiteConfigModule::SOURCE_DEFAULT},
+     *                        {@link SiteConfigModule::SOURCE_FILE},
+     *                        {@link SiteConfigModule::SOURCE_DATABASE},
+     *                        {@link SiteConfigModule::SOURCE_INSTANCE} or
+     *                        {@link SiteConfigModule::SOURCE_RUNTIME}
+     * @param mixed  $name
+     *
+     * @throws SiteException if there is no config setting defined for the
+     *                       given section and name
+     */
+    public function setSource($section, $name, $source)
+    {
+        if (!array_key_exists($section, $this->setting_sources)) {
+            throw new SiteException(sprintf(
+                "Section '%s' does not exist in this config module.",
+                $name
+            ));
+        }
+
+        $setting_names = $this->setting_sources[$section];
+
+        if (!array_key_exists($name, $this->setting_sources[$section])) {
+            throw new SiteException(sprintf(
+                "Setting '%s' does not exist in the section '%s'.",
+                $name,
+                $section
+            ));
+        }
+
+        $valid_sources = [self::SOURCE_DEFAULT, self::SOURCE_FILE, self::SOURCE_DATABASE, self::SOURCE_INSTANCE, self::SOURCE_RUNTIME];
+
+        if (!in_array($source, $valid_sources)) {
+            $source = self::SOURCE_RUNTIME;
+        }
+
+        $this->setting_sources[$section][$name] = $source;
+    }
+
+    /**
+     * Gets a section of this config module.
+     *
+     * @param string $name the name of the config section to get
+     *
+     * @return SiteConfigSection the config section
+     */
+    public function __get($name)
+    {
+        if (!array_key_exists($name, $this->sections)) {
+            throw new SiteException(sprintf(
+                "Section '%s' does not exist in this config module.",
+                $name
+            ));
+        }
+
+        return $this->sections[$name];
+    }
+
+    /**
+     * Checks for existence of a section within this config module.
+     *
+     * @param string $name the name of the section value to check
+     *
+     * @return bool true if the section exists in this config module and
+     *              false if it does not
+     */
+    public function __isset($name)
+    {
+        return array_key_exists($name, $this->sections);
+    }
+
+    /**
+     * Gets a string representation of this config module.
+     *
+     * This gets the loaded config values split into the loaded sections.
+     *
+     * @return string a string representation of this config module
+     */
+    public function __toString(): string
+    {
+        ob_start();
+        foreach ($this->sections as $section) {
+            echo $section, "\n";
+        }
+
+        return (string) ob_get_clean();
```

Comment:
Is there a better way to safely use ob_get_clean with phpstan? Can it be made to throw instead of return false?

Reply:
It does not appear that PHP has any way to get `ob_get_clean` to throw, unfortunately.  I suppose we could write our own wrapper for it, but I think that's probably overkill (and I don't know if it would help with analysis anyway, as it would likely only throw during execution).

There are a few other places in the code where we assume a string response from something that might return a string or non-string value.  I suppressed those in `rector.php` by skipping the `NullToStrictStringFuncCallArgRector` check.  As an example:

```diff
Site/pages/SiteContactPage.php:100
             // Filter bad character encoding. If invalid, assume ISO-8859-1
             // encoding and convert to UTF-8.
             if (!SwatString::validateUtf8($user_agent)) {
-                $user_agent = iconv('ISO-8859-1', 'UTF-8', $user_agent);
+                $user_agent = iconv('ISO-8859-1', 'UTF-8', (string) $user_agent);
             }

             // Only save if the user-agent was successfully converted to UTF-8.
             if ($user_agent !== false) {
                 // set max length based on database field length
-                $user_agent = mb_substr($user_agent, 0, 255);
+                $user_agent = mb_substr((string) $user_agent, 0, 255);
                 $message->user_agent = $user_agent;
             }
         }
```
`iconv()` returns string|false, so Rector wants to make sure it's actually a string when we pass it to `mb_substr()` (which it should be, because of the `if()` block it's in, so it might not look at the full context ?)

Reply:
Sounds good. I agree wrapping this is a bit overkill for now.